### PR TITLE
6176 event sourcing job processors - part 1

### PR DESCRIPTION
**Pull request by pihme**
## Description

Event sourcing modification for  :
* `JobIntent.CREATE/CREATED`
* `JobIntentCOMPLETE/COMPLETED`
* `JobIntent.FAIL/FAILED`

## Related issues

#6176 

The issue is not completed yet.

## Definition of Done

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by pihme** [comment]
I was unable to reproduce the test failure with the seeds given.

**Review by korthout — CHANGES_REQUESTED** [review]
Changes look good, but because some stuff is left for future work I'd prefer some extra markers. I also have a question about a test and some minor improvement requests. Please have another look

File: engine/src/main/java/io/zeebe/engine/processing/streamprocessor/CommandProcessor.java [hunk 1]
```diff
@@ -21,6 +23,15 @@ default boolean onCommand(final TypedRecord<T> command, final CommandControl<T>
     return true;
   }
 
+  // TODO clean up after refactoring; this is just a simple hook to be able to append additional
+  // commands/events
+  default void afterAccept(
+      final TypedCommandWriter commandWriter,
+      final StateWriter stateWriter,
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/streamprocessor/CommandProcessor.java:30]: Is your preference strong enough to request a change here?

File: engine/src/test/java/io/zeebe/engine/util/client/JobClient.java [hunk 2]
```diff
@@ -49,6 +49,7 @@ public JobClient(final StreamProcessorRule environmentRule) {
 
   public JobClient ofInstance(final long workflowInstanceKey) {
     this.workflowInstanceKey = workflowInstanceKey;
+    jobRecord.setWorkflowInstanceKey(workflowInstanceKey);
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/test/java/io/zeebe/engine/util/client/JobClient.java:52]: This was necessary to fix one of the tests. The test used this method to drive the engine. The test failed because the record  didn't have the right workflow instance key. I don't recall right now whether the error was in the assertion or the error occurred while running the tests and executing the logic. If you want to know more, I could remove the line again and see which test fails.

File: engine/src/main/java/io/zeebe/engine/processing/streamprocessor/CommandProcessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processing/streamprocessor/CommandProcessor.java:30]: This method is supposed to be used by all user commands from now on right? If so, then yes. If we plan on removing this later anyways, then I don't mind it

File: engine/src/test/java/io/zeebe/engine/util/client/JobClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/test/java/io/zeebe/engine/util/client/JobClient.java:52]: Since I don't understand why, I feel that it hides a potential bug. Please find out why the workflowinstance key isn't written on the record

File: engine/src/main/java/io/zeebe/engine/processing/streamprocessor/CommandProcessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/streamprocessor/CommandProcessor.java:30]: I think we will remove this later on. It is not used by all user commands so far. It is predominantly used by job related commands (however, there are exceptions). My plan was to refactor all job related processors to get a better feeling for this class. Afterwards, I want to refactor this class and also the impl class respectively. Either with the intention to remove the interface altogether or with the intention to align it closer to the other processors.

File: engine/src/main/java/io/zeebe/engine/processing/job/CompleteProcessor.java [hunk 3]
```diff
@@ -32,10 +42,40 @@ public boolean onCommand(
...
+
+    if (elementInstance != null) {
+      final long scopeKey = elementInstance.getValue().getFlowScopeKey();
+      final ElementInstance scopeInstance = elementInstanceState.getInstance(scopeKey);
+
+      if (scopeInstance != null && scopeInstance.isActive()) {
+        final WorkflowInstanceRecord workflowInstanceRecord = elementInstance.getValue();
+
+        // TODO send out COMPLETE_ELEMENT command when available, rename method; switch out for
+        // command writer
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/job/CompleteProcessor.java:75]: @saig0 Is there an issue to introduce the `COMPLETE_ELEMENT` command. I checked the list and didn't find any obvious candidate

File: engine/src/test/java/io/zeebe/engine/util/client/JobClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/test/java/io/zeebe/engine/util/client/JobClient.java:52]: Not sure exactly what you mean by "Please find out why the workflowinstance key isn't written on the record". How can I find out why something is not being done???

Anyways, I removed this line just to see whether the test might now pass due to other changes we made. It still fails. It fails in `JobIncidentTest.shouldCreateIncidentWithJobErrorMessage()`

This test writes a couple of commands using these lines:
```
// when
    ENGINE
        .job()
        .ofInstance(workflowInstanceKey)
        .withType(JOB_TYPE)
        .withErrorMessage("failed job")
        .fail();
```

And then looks for a record with those lines:
```
final Record failCommand =
        RecordingExporter.jobRecords()
            .withIntent(JobIntent.FAIL)
            .withWorkflowInstanceKey(workflowInstanceKey)
            .getFirst();
```

Which is where the test fails. There is a record with that intent, but it has -1 as workflow instance key.

The `ofInstance(..)` method is where I made the change. I don't know why it didn't set the workflow instance key before. To me it felt like an oversight. I don't see how consistent records can be written this way. So honestly, I didn't think much of it when I made the change. And I still see the change as an improvement.

But your question made me curious why it had worked in the past. And there was indeed a subtle difference:

```
    final Record failedEvent =
        RecordingExporter.jobRecords()
            .withIntent(JobIntent.FAILED)
            .withWorkflowInstanceKey(workflowInstanceKey)
            .getFirst();
```

So here it is looking for the event instead of the command. And processing being the way it is the event has the correct workflow instance key. 

So now when I undo my change in this class and I undo my change in `JobClient` the test passes again. I can only assume that it may have had to do with the problem of events not being processed if there was no command processor registered for the event.

So to take stock, I could undo both changes and then have a smaller change set.

I am a little bit stubborn here, though. I like the test based on the command better. I find it clearer. The way this is used later on is in this assertion:

```
    assertThat(incidentCommand.getSourceRecordPosition())
        .isEqualTo(failedEvent.getSourceRecordPosition());
```

Before the change it read:
```
assertThat(incidentCommand.getSourceRecordPosition()).isEqualTo(failedEvent.getPosition())
```

And if I would look for the command then it would read:
```
assertThat(incidentCommand.getSourceRecordPosition()).isEqualTo(failCommand.getPosition());
``` 

Which in my mind is easier to understand than `failedEvent.getSourceRecordPosition()`. And I also feel it is a more faithful adaptation of the test. The test checks - among others - that the incident has as source position the record that produced the incident. This used to be the FAILED event, and it now is the FAIL command. 

And I also like the change to `JobClient`. I find it immensely frustrating that these utility classes produce only half-consistent records. The job client already stores the workflow instance key internally and uses it in the `findJobkey()` method, but never uses it in writing any records. And all the other methods `withType(), withVariables(), withRetries()` they all modify the record. So this still feels like a simple oversight and an obvious improvement to make. 

So in summary. I would like to keep it as it is.

What do you think?

File: engine/src/main/java/io/zeebe/engine/processing/job/CompleteProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processing/job/CompleteProcessor.java:75]: Writing a [COMPLETE_ELEMENT](https://github.com/zeebe-io/zeebe/blob/develop/protocol/src/main/java/io/zeebe/protocol/record/intent/WorkflowInstanceIntent.java#L33) command should work :+1:

[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/job/CompleteProcessor.java:75]: It doesn't work yet, because there is no stream processor registered for the `COMPLETE_ELEMENT` command.

File: engine/src/test/java/io/zeebe/engine/util/client/JobClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/test/java/io/zeebe/engine/util/client/JobClient.java:52]: I had a chat with Phil and he explained to me the purpose of the `JobClient` class. With that information I will undo my changes

**Review by korthout — APPROVED** [review]
Thanks for the changes! LGTM 🚗  

EDIT: please also cleanup the commit history a bit when you've fixed the merge conflicts :)
